src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Providers/Providers/AzureFilesPsBackupBackupProvider.cs

```diff
@@ -57,7 +61,328 @@ public class AzureFilesPsBackupProvider : IPsBackupProvider
+            ValidateAzureFilesWorkloadType(policyWorkloadType);
+            if (itemWorkloadType != policyWorkloadType)
+            {
+                throw new ArgumentException(string.Format(Resources.UnExpectedWorkLoadTypeException,
+                                            CmdletModel.WorkloadType.AzureFiles.ToString(),
+                                            itemWorkloadType.ToString()));
+            }
+        }
+
+        private void ValidateFileShareEnableProtectionRequest(string fileShareName, string storageAccountName,
+            string storageAccountResourceGroupName, PolicyBase policy)
+        {
+            if (string.IsNullOrEmpty(fileShareName))
+            {
+                throw new ArgumentException(string.Format(Resources.InvalidFileShareName));
+            }
+            if (string.IsNullOrEmpty(storageAccountName) && string.IsNullOrEmpty(storageAccountResourceGroupName))
+            {
+                throw new ArgumentException(
+                    string.Format(Resources.BothStorageAccountNameAndStorageAccountResourceGroupNameShouldNotEmpty)
+                    );
+            }
+        }
+
+        private void ValidateAzureFilesModifyProtectionRequest(ItemBase itemBase,
+            PolicyBase policy)
+        {
+            if (itemBase == null || itemBase.GetType() != typeof(AzureFileShareItem))
+            {
+                throw new ArgumentException(string.Format(Resources.InvalidProtectionPolicyException,
+                                            typeof(AzureFileShareItem).ToString()));
+            }
+
+            if (string.IsNullOrEmpty(((AzureFileShareItem)itemBase).ParentContainerFabricId))
+            {
+                throw new ArgumentException(Resources.VirtualMachineIdIsEmptyOrNull);
+            }
+        }
+
+        private WorkloadProtectableItemResource GetAzureFileShareProtectableObject(
+            string azureFileShareName,
+            string storageAccountName,
+            string storageAccountResourceGroupName,
+            string vaultName = null,
+            string resourceGroupName = null)
+        {
+            //Trigger Discovery
+            ODataQuery<BMSRefreshContainersQueryObject> queryParam = new ODataQuery<BMSRefreshContainersQueryObject>(
+               q => q.BackupManagementType
+                    == ServiceClientModel.BackupManagementType.AzureStorage);
+            AzureWorkloadProviderHelper.RefreshContainer(vaultName, resourceGroupName, queryParam);
+
+            //get registered storage accounts
+            bool isRegistered = false;
+            string storageContainer = null;
+            string classicStorageContainer = null;
+            List<ContainerBase> registeredStorageAccounts = GetRegisteredStorageAccount(vaultName, resourceGroupName);
+            foreach (var storageAccount in registeredStorageAccounts)
+            {
+                string accountName = "Storage;" + resourceGroupName + ";" + storageAccountName;
```

Comment:
The vault resourceGroupName is being used here erroneously. Can you parse out the storageAccountName from the storageAccount.Name and check it against accountName using String.Compare?